src/api/client.ts

```diff
@@ -0,0 +1,197 @@
+/**
+ * RequestMethod
+ * The HTTP request methods that our http client supports.
+ */
+export type RequestMethod =
+  | "GET"
+  | "get"
+  | "PATCH"
+  | "patch"
+  | "PUT"
+  | "put"
+  | "POST"
+  | "post"
+  | "DELETE"
+  | "delete"
+
+/**
+ * RequestOptions
+ * The set of options available for any http client request.
+ */
+export interface RequestOptions {
+  /**
+   * Disable the full screen loading interface during the request when true.
+   */
+  skipLoader?: boolean
+  /**
+   * Artificially delay a request by the time specified when set.
+   */
+  withDelay?: number
+}
+
+export const HTTP_ERROR = "HttpError"
+export const HTTP_CANCELLED_ERROR = "HttpCanceledError"
+
+/**
+ * HttpError
+ * An http client error when the request is rejected.
+ */
+export class HttpError<T = unknown> extends Error {
+  /**
+   * The http response status code.
+   */
+  status: number
+
+  /**
+   * The http response body.
+   */
+  response?: T
+
+  constructor(message?: string, status?: number, response?: T, name?: string) {
+    super(message || "")
+    this.name = name || HTTP_ERROR
+    this.status = status || 0
+    this.response = response
+  }
+}
+
+/**
+ * HttpPromise
+ * The successfully resolved interface of an http client request.
+ */
+export interface HttpPromise<T = any> extends Promise<T> {}
+
+/**
+ * TrailsPromise
+ * The successfully resolved interface of an http client request returned by @xy-planning-network/Trails.
+ */
+export interface TrailsPromise<T = any> extends Promise<TrailsResponse<T>> {}
```

Comment:
The first of many convenience interfaces.  It saves you from the problem of deeply nested generics (feedback from portal implementations):

```ts
HttpPromise<TrailsResponse<User>>> === TrailsPromise<User>
```

Currently, the Tax app requires functions to annotate their return value, which is a typescript best practice, but is only covered by eslint and is not a Typescript rule.  Jury is out on whether that eslint rule should be used elsewhere.  Thee Tax app uses the `TrailsPromise` interface to annotate it's api calls.

Reply:
`HttpPromise` is effectively a copycat of `AxiosPromise`, but since trees doesn't have any kind of response schema and just forwards the response body, it ends up being a straight overlap of `Promise`.  I could see a day when it's time to wrap all responses in some kind of schema that would include the response code, headers, and data etc., which is why I've defined it early.  When/if it does happen, Tree's consumer can update, type check and know exactly where they need to update their access to the response body.

At the moment, the only use for `HttpPromise` would be if we need to hit a 3rd party API endpoint.